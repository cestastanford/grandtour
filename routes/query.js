/*
*   Imports
*/

const router = require('express').Router()
const { isViewer } = require('./auth')
const Entry = require('../models/entry')
const entries = require('../controllers/entries')


/*
*   Retrieves the counts of entries that have values for the queries
*   defined in './mappings/counts".
*/

router.get('/api/getcount', isViewer, (req, res, next) => {

    entries.getCounts(req.user.activeRevisionIndex)
    .then(counts => res.json(counts))
    .catch(next)

})


/*
*   Previously-existing entry search routes.
*/

router.post('/api/entries/search', isViewer, entries.search);
router.post('/api/entries/search2', isViewer, entries.search2);
router.post('/api/entries/suggest', isViewer, entries.suggest);
router.post('/api/entries/export', isViewer, entries.export);
router.post('/api/entries/uniques', isViewer, entries.uniques);


/*
*   Exports
*/

module.exports = router
