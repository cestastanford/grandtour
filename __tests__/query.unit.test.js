import { parseQuery } from "../query";

describe('test parseQuery', () => {
    test('empty query', () => {
        const query = {
        };
        const result = {
        };
        expect(parseQuery(query)).toEqual(result);
    });
    test('compound query', () => {
        const query = {
            "occupations_group": ["Diplomacy", "Clergy"], "pursuits": ["diplomat"]
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test('query with object', () => {
        const query = {
            "occupations_group": [{ "_id": "Diplomacy" }]
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test('single query with string - should give fuzzy search', () => {
        const query = {
            "occupations_group": "Diplomacy"
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test('negative query', () => {
        const query = {
            "occupations_group": [{ "_id": "Diplomacy", "negative": true }, "Clergy"], "pursuits": ["diplomat"]
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test('empty query with OR', () => {
        const query = {
            "pursuits": { "operator": "or", "uniques": [] }
        };
        const result = {
        };
        expect(parseQuery(query)).toEqual(result);
    });
    test('query OR with object', () => {
        const query = {
            "occupations_group": { "operator": "or", "uniques": [{ "_id": "Diplomacy" }] }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test('query AND with object', () => {
        const query = {
            "occupations_group": { "operator": "and", "uniques": [{ "_id": "Diplomacy" }, { "_id": "Democracy" }] }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test('negative OR query', () => {
        const query = {
            "occupations_group": { "operator": "or", "uniques": [{ "_id": "Diplomacy", "negative": true }, { "_id": "Clergy" }] },
            "pursuits": { "operator": "or", "uniques": [{ "_id": "diplomat" }] }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test('negative array (fullName and alternateNames) query', () => {
        const query = {
            "fullName": { "operator": "and", "uniques": [{ "_id": "John Taylor", "negative": true }] },
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
});

describe('test parseQuery with freeSearch', () => {
    test('regular freeSearch query', () => {
        const query = {
            "entry": {
                "terms": [
                    {
                        "value": "two",
                        "beginning": true,
                        "end": true,
                        "negative": false
                    },
                    {
                        "value": "three",
                        "beginning": true,
                        "end": true,
                        "negative": false
                    }
                ],
                "sections": [
                    {
                        "key": "biography",
                        "name": "Biography",
                        "checked": true
                    },
                    {
                        "key": "narrative",
                        "name": "Narrative",
                        "checked": true
                    },
                    {
                        "key": "tours",
                        "name": "Tours",
                        "checked": true
                    },
                    {
                        "key": "notes",
                        "name": "Notes",
                        "checked": true
                    }
                ],
                "operator": "or"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });

    test('freeSearch query with "and" operator', () => {
        const query = {
            "entry": {
                "terms": [
                    {
                        "value": "two",
                        "beginning": true,
                        "end": true,
                        "negative": false
                    },
                    {
                        "value": "three",
                        "beginning": true,
                        "end": true,
                        "negative": false
                    }
                ],
                "sections": [
                    {
                        "key": "biography",
                        "name": "Biography",
                        "checked": true
                    },
                    {
                        "key": "narrative",
                        "name": "Narrative",
                        "checked": true
                    },
                    {
                        "key": "tours",
                        "name": "Tours",
                        "checked": true
                    },
                    {
                        "key": "notes",
                        "name": "Notes",
                        "checked": true
                    }
                ],
                "operator": "and"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });

    test('freeSearch query with negatives', () => {
        const query = {
            "entry": {
                "terms": [
                    {
                        "value": "two",
                        "beginning": true,
                        "end": true,
                        "negative": true
                    },
                    {
                        "value": "three",
                        "beginning": true,
                        "end": true
                    }
                ],
                "sections": [
                    {
                        "key": "biography",
                        "name": "Biography",
                        "checked": true
                    },
                    {
                        "key": "narrative",
                        "name": "Narrative",
                        "checked": true
                    },
                    {
                        "key": "tours",
                        "name": "Tours",
                        "checked": true
                    },
                    {
                        "key": "notes",
                        "name": "Notes",
                        "checked": true
                    }
                ]
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
});

describe("test parseQuery with dates", () => {
    test("month range", () => {
        const query = {
            travelDate: {
                startMonth: "1",
                endMonth: "1"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("single month", () => {
        const query = {
            travelDate: {
                startMonth: "1"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("single year", () => {
        const query = {
            travelDate: {
                startYear: "1700"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("single year and single month", () => {
        const query = {
            travelDate: {
                startYear: "1700",
                startMonth: "5"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("year range", () => {
        const query = {
            travelDate: {
                startYear: "1700",
                endYear: "1780"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("month range + year range", () => {
        const query = {
            travelDate: {
                startYear: "1700",
                endYear: "1780",
                startMonth: "1",
                endMonth: "10"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("month range + single year", () => {
        const query = {
            travelDate: {
                startYear: "1700",
                startMonth: "1",
                endMonth: "10"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    describe("repeated years/months should be treated the same as a single year/month", () => {
        describe("startYear", () => {

            expect(parseQuery({
                travelDate: {
                    startYear: "1700",
                }
            })).toEqual(parseQuery({
                travelDate: {
                    startYear: "1700",
                    endYear: "1700"
                }
            }));
        });
        describe("startMonth", () => {
            expect(parseQuery({
                travelDate: {
                    startMonth: "5"
                }
            })).toEqual(parseQuery({
                travelDate: {
                    startMonth: "5",
                    endMonth: "5"
                }
            }));
        });
        describe("startYear + startMonth", () => {
            expect(parseQuery({
                travelDate: {
                    startYear: "1700",
                    startMonth: "5",
                }
            })).toEqual(parseQuery({
                travelDate: {
                    startYear: "1700",
                    endYear: "1700",
                    startMonth: "5",
                    endMonth: "5"
                }
            }));
        });
    });
    test("date + place", () => {
        const query = {
            travelDate: {
                startYear: "1700",
                startMonth: "1",
                endMonth: "10"
            },
            travelPlace: {"operator":"and","uniques":[{"negative":false,"_id":"Genoa"}]}
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("date + negative place", () => {
        const query = {
            travelDate: {
                startYear: "1700",
                startMonth: "1",
                endMonth: "10"
            },
            travelPlace: {"operator":"and","uniques":[{"negative":true,"_id":"Genoa"}]}
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
});

describe("test parseQuery with index", () => {
    test("single index", () => {
        const query = {
            index: {
                startIndex: "1",
                endIndex: "1"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("multiple indices", () => {
        const query = {
            index: {
                startIndex: "1,2,3",
                endIndex: "1,2,3"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("range", () => {
        const query = {
            index: {
                startIndex: "1-3",
                endIndex: "1-3"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("multiple indices + range", () => {
        const query = {
            index: {
                startIndex: "1,2,3, 4-6",
                endIndex: "1,2,3, 4-6"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
    test("different start and end with multiple indices + range", () => {
        const query = {
            index: {
                startIndex: "1,2,3,4-6",
                endIndex: "10-12,13,14,15"
            }
        };
        expect(parseQuery(query)).toMatchSnapshot();
    });
});