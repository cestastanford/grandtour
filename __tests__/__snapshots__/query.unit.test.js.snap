// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test parseQuery compound query 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "occupations": Object {
            "$elemMatch": Object {
              "group": "Diplomacy",
            },
          },
        },
        Object {
          "occupations": Object {
            "$elemMatch": Object {
              "group": "Clergy",
            },
          },
        },
      ],
    },
    Object {
      "$or": Array [
        Object {
          "pursuits": Object {
            "$elemMatch": Object {
              "pursuit": "diplomat",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`test parseQuery negative OR query 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "occupations": Object {
            "$ne": Object {
              "$elemMatch": Object {
                "group": "Diplomacy",
              },
            },
          },
        },
        Object {
          "occupations": Object {
            "$elemMatch": Object {
              "group": "Clergy",
            },
          },
        },
      ],
    },
    Object {
      "$or": Array [
        Object {
          "pursuits": Object {
            "$elemMatch": Object {
              "pursuit": "diplomat",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`test parseQuery negative query 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "occupations": Object {
            "$ne": Object {
              "$elemMatch": Object {
                "group": "Diplomacy",
              },
            },
          },
        },
        Object {
          "occupations": Object {
            "$elemMatch": Object {
              "group": "Clergy",
            },
          },
        },
      ],
    },
    Object {
      "$or": Array [
        Object {
          "pursuits": Object {
            "$elemMatch": Object {
              "pursuit": "diplomat",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`test parseQuery query AND with object 1`] = `
Object {
  "$and": Array [
    Object {
      "$and": Array [
        Object {
          "occupations": Object {
            "$elemMatch": Object {
              "group": "Diplomacy",
            },
          },
        },
        Object {
          "occupations": Object {
            "$elemMatch": Object {
              "group": "Democracy",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`test parseQuery query OR with object 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "occupations": Object {
            "$elemMatch": Object {
              "group": "Diplomacy",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`test parseQuery query with object 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "occupations": Object {
            "$elemMatch": Object {
              "group": "Diplomacy",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`test parseQuery single query with string - should give fuzzy search 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "occupations": Object {
            "$elemMatch": Object {
              "group": Object {
                "$regex": /Diplomacy/gi,
              },
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`test parseQuery with dates month range + single year 1`] = `
Object {
  "$and": Array [
    Object {
      "travels": Object {
        "$elemMatch": Object {
          "$or": Array [
            Object {
              "$and": Array [
                Object {
                  "$and": Array [
                    Object {
                      "travelEndYear": Object {
                        "$gte": 1700,
                      },
                    },
                    Object {
                      "travelStartYear": Object {
                        "$lte": 1700,
                      },
                    },
                  ],
                },
                Object {
                  "$or": Array [
                    Object {
                      "travelStartMonth": Object {
                        "$exists": false,
                      },
                    },
                    Object {
                      "travelEndMonth": Object {
                        "$exists": false,
                      },
                    },
                    Object {
                      "travelStartMonth": Object {
                        "$eq": 0,
                      },
                    },
                    Object {
                      "travelEndMonth": Object {
                        "$eq": 0,
                      },
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartMonth": Object {
                            "$gte": 1,
                          },
                        },
                        Object {
                          "travelEndMonth": Object {
                            "$lte": 10,
                          },
                        },
                      ],
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartMonth": Object {
                            "$lte": 1,
                          },
                        },
                        Object {
                          "travelEndMonth": Object {
                            "$gte": 1,
                          },
                        },
                      ],
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartMonth": Object {
                            "$lte": 10,
                          },
                        },
                        Object {
                          "travelEndMonth": Object {
                            "$gte": 10,
                          },
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`test parseQuery with dates month range + year range 1`] = `
Object {
  "$and": Array [
    Object {
      "travels": Object {
        "$elemMatch": Object {
          "$or": Array [
            Object {
              "$and": Array [
                Object {
                  "$or": Array [
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartYear": Object {
                            "$gte": 1700,
                          },
                        },
                        Object {
                          "travelEndYear": Object {
                            "$lte": 1780,
                          },
                        },
                      ],
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartYear": Object {
                            "$lte": 1700,
                          },
                        },
                        Object {
                          "travelEndYear": Object {
                            "$gte": 1700,
                          },
                        },
                      ],
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartYear": Object {
                            "$lte": 1780,
                          },
                        },
                        Object {
                          "travelEndYear": Object {
                            "$gte": 1780,
                          },
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "$or": Array [
                    Object {
                      "travelStartMonth": Object {
                        "$exists": false,
                      },
                    },
                    Object {
                      "travelEndMonth": Object {
                        "$exists": false,
                      },
                    },
                    Object {
                      "travelStartMonth": Object {
                        "$eq": 0,
                      },
                    },
                    Object {
                      "travelEndMonth": Object {
                        "$eq": 0,
                      },
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartMonth": Object {
                            "$gte": 1,
                          },
                        },
                        Object {
                          "travelEndMonth": Object {
                            "$lte": 10,
                          },
                        },
                      ],
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartMonth": Object {
                            "$lte": 1,
                          },
                        },
                        Object {
                          "travelEndMonth": Object {
                            "$gte": 1,
                          },
                        },
                      ],
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartMonth": Object {
                            "$lte": 10,
                          },
                        },
                        Object {
                          "travelEndMonth": Object {
                            "$gte": 10,
                          },
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`test parseQuery with dates month range 1`] = `
Object {
  "$and": Array [
    Object {
      "travels": Object {
        "$elemMatch": Object {
          "$or": Array [
            Object {
              "$and": Array [
                Object {
                  "$or": Array [
                    Object {
                      "travelStartMonth": Object {
                        "$exists": false,
                      },
                    },
                    Object {
                      "travelEndMonth": Object {
                        "$exists": false,
                      },
                    },
                    Object {
                      "travelStartMonth": Object {
                        "$eq": 0,
                      },
                    },
                    Object {
                      "travelEndMonth": Object {
                        "$eq": 0,
                      },
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelEndMonth": Object {
                            "$gte": 1,
                          },
                        },
                        Object {
                          "travelStartMonth": Object {
                            "$lte": 1,
                          },
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`test parseQuery with dates single month 1`] = `
Object {
  "$and": Array [
    Object {
      "travels": Object {
        "$elemMatch": Object {
          "$or": Array [
            Object {
              "$and": Array [
                Object {
                  "$or": Array [
                    Object {
                      "travelStartMonth": Object {
                        "$exists": false,
                      },
                    },
                    Object {
                      "travelEndMonth": Object {
                        "$exists": false,
                      },
                    },
                    Object {
                      "travelStartMonth": Object {
                        "$eq": 0,
                      },
                    },
                    Object {
                      "travelEndMonth": Object {
                        "$eq": 0,
                      },
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelEndMonth": Object {
                            "$gte": 1,
                          },
                        },
                        Object {
                          "travelStartMonth": Object {
                            "$lte": 1,
                          },
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`test parseQuery with dates single year 1`] = `
Object {
  "$and": Array [
    Object {
      "travels": Object {
        "$elemMatch": Object {
          "$or": Array [
            Object {
              "$and": Array [
                Object {
                  "$and": Array [
                    Object {
                      "travelEndYear": Object {
                        "$gte": 1700,
                      },
                    },
                    Object {
                      "travelStartYear": Object {
                        "$lte": 1700,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`test parseQuery with dates single year and single month 1`] = `
Object {
  "$and": Array [
    Object {
      "travels": Object {
        "$elemMatch": Object {
          "$or": Array [
            Object {
              "$and": Array [
                Object {
                  "$and": Array [
                    Object {
                      "travelEndYear": Object {
                        "$gte": 1700,
                      },
                    },
                    Object {
                      "travelStartYear": Object {
                        "$lte": 1700,
                      },
                    },
                  ],
                },
                Object {
                  "$or": Array [
                    Object {
                      "travelStartMonth": Object {
                        "$exists": false,
                      },
                    },
                    Object {
                      "travelEndMonth": Object {
                        "$exists": false,
                      },
                    },
                    Object {
                      "travelStartMonth": Object {
                        "$eq": 0,
                      },
                    },
                    Object {
                      "travelEndMonth": Object {
                        "$eq": 0,
                      },
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelEndMonth": Object {
                            "$gte": 5,
                          },
                        },
                        Object {
                          "travelStartMonth": Object {
                            "$lte": 5,
                          },
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`test parseQuery with dates year range 1`] = `
Object {
  "$and": Array [
    Object {
      "travels": Object {
        "$elemMatch": Object {
          "$or": Array [
            Object {
              "$and": Array [
                Object {
                  "$or": Array [
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartYear": Object {
                            "$gte": 1700,
                          },
                        },
                        Object {
                          "travelEndYear": Object {
                            "$lte": 1780,
                          },
                        },
                      ],
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartYear": Object {
                            "$lte": 1700,
                          },
                        },
                        Object {
                          "travelEndYear": Object {
                            "$gte": 1700,
                          },
                        },
                      ],
                    },
                    Object {
                      "$and": Array [
                        Object {
                          "travelStartYear": Object {
                            "$lte": 1780,
                          },
                        },
                        Object {
                          "travelEndYear": Object {
                            "$gte": 1780,
                          },
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`test parseQuery with freeSearch freeSearch query with "and" operator 1`] = `
Object {
  "$and": Array [
    Object {
      "$and": Array [
        Object {
          "$or": Array [
            Object {
              "biography": Object {
                "$regex": /\\\\btwo\\\\b/gi,
              },
            },
            Object {
              "narrative": Object {
                "$regex": /\\\\btwo\\\\b/gi,
              },
            },
            Object {
              "tours": Object {
                "$regex": /\\\\btwo\\\\b/gi,
              },
            },
            Object {
              "notes": Object {
                "$regex": /\\\\btwo\\\\b/gi,
              },
            },
          ],
        },
        Object {
          "$or": Array [
            Object {
              "biography": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
            Object {
              "narrative": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
            Object {
              "tours": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
            Object {
              "notes": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`test parseQuery with freeSearch freeSearch query with negatives 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "$and": Array [
            Object {
              "biography": Object {
                "$regex": /\\^\\(\\(\\?!\\\\btwo\\\\b\\)\\.\\)\\*\\$/gi,
              },
            },
            Object {
              "narrative": Object {
                "$regex": /\\^\\(\\(\\?!\\\\btwo\\\\b\\)\\.\\)\\*\\$/gi,
              },
            },
            Object {
              "tours": Object {
                "$regex": /\\^\\(\\(\\?!\\\\btwo\\\\b\\)\\.\\)\\*\\$/gi,
              },
            },
            Object {
              "notes": Object {
                "$regex": /\\^\\(\\(\\?!\\\\btwo\\\\b\\)\\.\\)\\*\\$/gi,
              },
            },
          ],
        },
        Object {
          "$or": Array [
            Object {
              "biography": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
            Object {
              "narrative": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
            Object {
              "tours": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
            Object {
              "notes": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`test parseQuery with freeSearch regular freeSearch query 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "$or": Array [
            Object {
              "biography": Object {
                "$regex": /\\\\btwo\\\\b/gi,
              },
            },
            Object {
              "narrative": Object {
                "$regex": /\\\\btwo\\\\b/gi,
              },
            },
            Object {
              "tours": Object {
                "$regex": /\\\\btwo\\\\b/gi,
              },
            },
            Object {
              "notes": Object {
                "$regex": /\\\\btwo\\\\b/gi,
              },
            },
          ],
        },
        Object {
          "$or": Array [
            Object {
              "biography": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
            Object {
              "narrative": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
            Object {
              "tours": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
            Object {
              "notes": Object {
                "$regex": /\\\\bthree\\\\b/gi,
              },
            },
          ],
        },
      ],
    },
  ],
}
`;
